### DISCUSSION TITLE
How to own a part of array, just like what OpenCV does?

### DISCUSSION THREAD

**Discussion by fzyzcjy** [rust-ndarray/ndarray#1099] [Q&A]
Hi thanks for the lib! I want to own a part of array, just like what OpenCV does. Consider the (simplified) code:

```
fn read_and_crop_image() -> ??? {
  let big_image: Array2<u8> = read_the_image();
  let cropped = big_image.slice(s![0..3000, 0..4000]);
  cropped // ERROR!!
}
```

However it errors, since ArrayView should not outlive the Array it belongs to. I have tried `ArcArray` but it does not seem to support `.slice()` as well.

With OpenCV I can do:

```
fn read_and_crop_image() -> Mat {
  let big_image: Mat = read_the_image();
  let cropped = Mat::roi([0..3000, 0..4000], big_image);
  cropped // no problem, it is another Mat, and underlying it is reference counting
}
```

**Comment by jturner314** [comment]
You could use [`.slice_move()`](https://docs.rs/ndarray/0.15.3/ndarray/struct.ArrayBase.html#method.slice_move), [`.slice_collapse()`](https://docs.rs/ndarray/0.15.3/ndarray/struct.ArrayBase.html#method.slice_collapse), or other similar methods, and return an owned `Array2<u8>`, e.g.:

```rust
fn read_and_crop_image() -> Array2<u8> {
    let big_image: Array2<u8> = read_the_image();
    let cropped = big_image.slice_move(s![0..3000, 0..4000]);
    cropped
}
```

You could also do something similar with `ArcArray`. The key is to return an owned array type (`Array` or `ArcArray`), rather than a view of an array which will be dropped at the end of the function body.

Another option is to copy the relevant portion, which may be beneficial if the original image is much larger than the cropped portion:

```rust
fn read_and_crop_image() -> Array2<u8> {
    let big_image: Array2<u8> = read_the_image();
    let cropped = big_image.slice(s![0..3000, 0..4000]).to_owned();
    cropped
}
```

**Comment by fzyzcjy** [comment]
Thank you!
